### DISCUSSION TITLE
How to solve out of memory error while fitting the model?

### DISCUSSION THREAD

**Discussion by prateekgml** [microsoft/FLAML#579] [Q&A]
Hi ,

Can you please help me to resolve out of memory error while fitting the model-

![image](https://user-images.githubusercontent.com/97145738/172767334-210ccd62-b293-4719-bc72-07c80f976134.png)

**Comment by prateekgml** [comment]
If I take a fraction of the data for example 50% of data then I am able to proceed but getting the following error now:

![image](https://user-images.githubusercontent.com/97145738/172843707-ef8f062f-b7cf-4e33-b81f-2ef67c77cffb.png)

So it will be helpful if anybody can tell how to fit the model using whole data and about the above error.

**Reply by sonichi** [reply]
Do you get OOM error when using catboost directly?

**Reply by prateekgml** [reply]
@sonichi FIrst I used whole data like and I also used `use_ray=True` argument while fitting the model but I faced the same OOM error.

Then I reduced the dataset to 55% and then I was able to fit the model but now `CatBoostError: catboost/private/libs/target/target_converter.cpp:375: Target contains only one unique value` error is showing.

**Reply by prateekgml** [reply]
@sonichi any update on this OOM error?

As a next experiment, I tried to fit the model with only 50% data but with the estimator list, this time and again I faced the OOM error. Please see this trace from the console after that OOM appeared-

![image](https://user-images.githubusercontent.com/97145738/173171567-bd6b04c2-d0c6-4931-9e2d-ac19e15e70f7.png)

It seems this is a genuine issue with FLAML. Let me know if you need anything from my side to fix this issue.

**Reply by sonichi** [reply]
@prateekgml Does OOM error happen when you use catboost directly without using FLAML?
* If the answer is yes, then it's not an issue with FLAML.
* If the answer is no, does the OOM error happen when you set `use_ray=False`?

**Reply by prateekgml** [reply]
@sonichi This OOM occurred whether I enable use_ray or not and whether I use 100% or 50% data with FLAML.
If I use Catboost or LightGBM or XGBoost directly, this OOM error does not appear.

**Reply by sonichi** [reply]
> @sonichi This OOM occurred whether I enable use_ray or not and whether I use 100% or 50% data with FLAML. If I use Catboost or LightGBM or XGBoost directly, this OOM error does not appear.

Could you share the Kaggle notebook which contains these experiments? My account is sonichi

**Reply by prateekgml** [reply]
I have shared my two notebooks with you on Kaggle. Please check.

**Reply by sonichi** [reply]
I see that no `time_budget` was provided to `AutoML.fit`. In that case it will just train one configuration per model (using our [zero-shot](https://microsoft.github.io/FLAML/docs/Use-Cases/Zero-Shot-AutoML) technique). Is that intended? It's possible that the zero-shot configuration on your very large training data caused the OOM error. The OOM error hasn't happened in my run of the amex_perquet notebook though after 10 minutes. It's still training the first lgbm model.
<img width="743" alt="image" src="https://user-images.githubusercontent.com/4250911/173196425-2a81d576-450f-4ca2-9445-98f9a73f8d04.png">

If you assign a `time_budget` in `AutoML.fit`, FLAML will perform hyperparameter search and begin with small models and subsamples, which should not cause OOM error in the beginning.

**Reply by prateekgml** [reply]
@sonichi any recommended value for `time_budget`? I set as `time_budget=600` and I tried to fit the whole dataset as `automl.fit(X_train=X_train, y_train=y_train, task='classification', estimator_list=["lgbm"], time_budget=600, ensemble=False, use_ray=True, seed=7654321, log_file_name="flaml_logs.log",max_iter=1)` 

But it gave me OOM error. Also please confirm if it is using GPU or explicitly I need to enable GPU training because in below screenshot circled in blue,  GPU is not used-

![image](https://user-images.githubusercontent.com/97145738/173219498-fada3c34-5e37-48d9-92e7-c339dffef1d3.png)

**Reply by sonichi** [reply]
@prateekgml I tried `lighgbm.LGBMClassifier.fit` without flaml, it gets OOM error too on the whole dataset. So it's not an FLAML-specific issue. Your dataset is too large to fit in RAM of this VM.
FLAML supports GPU for XGBoost. To use that, please limit the estimator_list to be ["xgboost", "xgb_limitdepth"] and specify `gpu_per_trial=1` in `AutoML.fit()`.
BTW, when you use "max_iter=1", it will only train one model. For a general guide about how to set time_budget, check https://microsoft.github.io/FLAML/docs/Use-Cases/Task-Oriented-AutoML#how-to-set-time-budget

**Reply by prateekgml** [reply]
@sonichi Thanks!

But I can see many individuals' notebooks using catboost or lightgbm or xgboost and they were able to train their models. A few examples are as below-

[https://www.kaggle.com/competitions/amex-default-prediction/code](url)

[https://www.kaggle.com/code/karandora/xgboost-optuna-baseline](url)

Meanwhile, I will continue with my experiments and let you know if I succeed with FLAML.

**Reply by sonichi** [reply]
I can't access these notebooks. Have you ever made one successful run with catboost/lightgbm/xgboost in the same environment?

**Reply by prateekgml** [reply]
@sonichi I have done my experiment using XGBoost directly to the same dataset and the entire training with 5 fold has been done in 12 minutes. You can check 15the version of my notebook that I had shared with you-
[https://www.kaggle.com/code/iamprateek/amex-perquet?scriptVersionId=98308716](url)

**Reply by sonichi** [reply]
In that case, you can use the [tune user-defined function](https://microsoft.github.io/FLAML/docs/Getting-Started#tune-user-defined-function) API to do hyperparameter search. As long as the training for each configuration can succeed, the tuning would not cause OOM.
